pkg/events/derive/net_tcp.go

```diff
@@ -0,0 +1,106 @@
+package derive
+
+import (
+	"errors"
+	"strconv"
+
+	"github.com/aquasecurity/tracee/pkg/dnscache"
+	"github.com/aquasecurity/tracee/pkg/errfmt"
+	"github.com/aquasecurity/tracee/pkg/events"
+	"github.com/aquasecurity/tracee/pkg/events/parse"
+	"github.com/aquasecurity/tracee/pkg/logger"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+func NetTCPConnect(cache *dnscache.DNSCache) DeriveFunction {
+	return deriveSingleEvent(events.NetTCPConnect, deriveNetTCPConnectArgs(cache))
+}
+
+func deriveNetTCPConnectArgs(cache *dnscache.DNSCache) deriveArgsFunction {
+	return func(event trace.Event) ([]interface{}, error) {
+		dstIP, dstPort, err := pickIpAndPort(event, "remote_addr")
+		if err != nil {
+			logger.Debugw("error picking address", "error", err)
+			return nil, nil
+		}
+		if dstIP == "" {
+			return nil, nil
+		}
+
+		// Check if DNS cache is enabled
+		results := []string{}
+		if cache != nil {
+			query, err := cache.Get(dstIP)
+			if err != nil {
+				switch err {
+				case dnscache.ErrDNSRecordNotFound, dnscache.ErrDNSRecordExpired:
+					results = []string{}
+					goto allGood
+				}
+				logger.Debugw("ip lookup error", "ip", dstIP, "error", err)
+				return nil, nil
+			}
+			results = query.DNSResults()
+		}
+
+	allGood:
+		return []interface{}{
+			dstIP,
+			dstPort,
+			results,
+		}, nil
+	}
+}
+
+// pickIpAndPort returns the IP address and port from the event's sockaddr field.
+func pickIpAndPort(event trace.Event, fieldName string) (string, int, error) {
+	var err error
+	// e.g: sockaddr: map[sa_family:AF_INET sin_addr:10.10.11.2 sin_port:1234]
+
+	sockaddr, err := parse.ArgVal[map[string]string](event.Args, fieldName)
+	if err != nil {
+		return "", 0, errfmt.WrapError(err)
+	}
+	if sockaddr == nil {
+		return "", 0, errfmt.WrapError(errors.New("remote_addr not found"))
+	}
+
+	// getAddrAndPort returns the address and port from sockaddr for IPv4 or IPv6.
+	getAddrAndPort := func(a, p string, sockaddr map[string]string) (string, int, error) {
```

Comment:
Agreed! I wonder if compiler is smart to optimize this on its own, but I'll move the func to make sure.

Reply:
I've seen instances of this allocation in pprof analyses, so if it is smart enough to do so, it's not at least not guaranteed. But it also has been a few go version since then.